### PR TITLE
UI Rework + Docker Support

### PR DESCRIPTION
**Pull request by vanUbor**
## Summary by CodeRabbit

- **Neue Funktionen**
  - Futuristisches, cyberpunk-inspiriertes Benutzerinterface: Das Kontrollpanel und die Statusanzeigen wurden modernisiert, mit interaktiven Animationen, akustischem Feedback und neu gestalteten Schaltflächen für ein verbessertes Nutzererlebnis.
  - Verbesserte Build- und Sicherheitskonfiguration, die eine stabilere und sicherere Anwendung gewährleistet.
  - Einführung eines neuen Docker-Setups zur Vereinfachung der Entwicklungsumgebung.
  
- **Dokumentation**
  - Ergänzende Entwicklungsrichtlinien und Bedienhinweise wurden hinzugefügt.
  - Neue Dokumentation zu Docker-Befehlen zur Verwaltung der Projektumgebung.

- **Stil & Design**
  - Überarbeitete Farben, Schriften und Animationseffekte verleihen der Anwendung ein modernes, ansprechendes Erscheinungsbild.

### CODE REVIEW

**Comment by vanUbor** [comment]
@coderabbitai review

File: WireMock/wwwroot/css/site-cyberpunk.css [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+  border-color: var(--neon-primary);
+  color: var(--neon-primary);
+  box-shadow: 0 0 10px rgba(0, 238, 255, 0.3);
+}
+
+.page-item.disabled .page-link {
+  background-color: transparent;
+  border-color: var(--cyber-border);
+  color: var(--cyber-text-secondary);
+}
```
[the remark this answers is not in the document]
Comment (vanUbor) [reply WireMock/wwwroot/css/site-cyberpunk.css]: should be fixed, please verify

File: WireMock/wwwroot/css/site-cyberpunk.css [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+
+textarea::-webkit-scrollbar-thumb {
+  background: var(--cyber-border);
+  border-radius: 4px;
+}
+
+textarea::-webkit-scrollbar-thumb:hover {
+  background: var(--neon-primary);
+}
```
[the remark this answers is not in the document]
Comment (vanUbor) [reply WireMock/wwwroot/css/site-cyberpunk.css:478]: I fixed it, please verify

File: WireMock/wwwroot/css/site-cyberpunk.css (hunk 1 above)
[the remark this answers is not in the document]
Comment (vanUbor) [reply WireMock/wwwroot/css/site-cyberpunk.css]: ich habe deine zusätzliche Anmerkung berücksichtigt und erneut änderungen durchgeführt, bitte überprüfe ob ich sowohl deine originalen als auch deine zusätzlichen Kommentare zu deiner zufriedenheit umgesetzt habe

File: WireMock/wwwroot/css/site-cyberpunk.css (hunk 2 above)
[the remark this answers is not in the document]
Comment (vanUbor) [reply WireMock/wwwroot/css/site-cyberpunk.css:470]: jetzt sollte es teil der css sein, bitte überprüfe nochmals

File: WireMock/wwwroot/scripts/MatcherUI/HeaderMatcherUI.ts [hunk 3]
```diff
@@ -37,12 +38,18 @@ export default class HeaderMatcherUI extends MatcherUI {
...
                 rawMapContent.Request.Headers[headerIndex].Matchers.splice(matcherIndex, 1);
-                rawMap.textContent = JSON.stringify(rawMapContent, null, 1);
-                matcherRow.remove();
+                if (rawMapContent.Request.Headers[headerIndex].Matchers.length > 0) {
+                    rawMap.textContent = JSON.stringify(rawMapContent, null, 1);
+                    matcherRow.remove();
+                } else {
+                    this.showErrorModal("You cannot remove the last matcher - " +
+                        "Remove the matcher collection instead");
+                }
```
[the remark this answers is not in the document]
Comment (vanUbor) [reply WireMock/wwwroot/scripts/MatcherUI/HeaderMatcherUI.ts]: sollte gelöst sein, bitte prüfe den fix

File: WireMock/wwwroot/scripts/MatcherUI/PathMatcherUI.ts [hunk 4]
```diff
@@ -47,8 +47,14 @@ export default class PathMatcherUI extends MatcherUI {
...
-                    rawMap.textContent = JSON.stringify(rawMapContent, null, 1);
-                    matcherRow.remove();
+                    if (rawMapContent.Request.Path.Matchers.length > 0) {
+                        rawMap.textContent = JSON.stringify(rawMapContent, null, 1);
+                        matcherRow.remove();
+                    }
+                    else {
+                        this.showErrorModal("You cannot remove the last matcher - " +
+                            "Paths requires at least one matcher");
+                    }
```
[the remark this answers is not in the document]
Comment (vanUbor) [reply WireMock/wwwroot/scripts/MatcherUI/PathMatcherUI.ts]: sollte gelöst sein, bitte prüfe den fix

File: WireMock/wwwroot/scripts/helper/createElement.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    function setAttributeIfApply() {
+        if (options.attributes) {
+            for (const key in options.attributes) {
+                if (options.attributes.hasOwnProperty(key)) {
+                    element.setAttribute(key, options.attributes[key]);
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (vanUbor) [reply WireMock/wwwroot/scripts/helper/createElement.ts]: sollte gelöst sein, bitte prüfe den fix, ich habe nicht deine vorgeschlagene Änderung vorgenommen sondern nutze Object.entries() stattdessen.

File: WireMock/wwwroot/scripts/helper/createElement.ts [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        if (options.innerText !== undefined) {
+            element.innerText = options.innerText;
+        }
+    }
+
+    setOptionIfApply();
+    setAttributeIfApply();
+    setStyleIfApply();
+    setInnerHtmlIfApply();
+    setInnerTextIfApplly();
```
[the remark this answers is not in the document]
Comment (vanUbor) [reply WireMock/wwwroot/scripts/helper/createElement.ts]: sollte gelöst sein, bitte prüfe den fix

File: WireMock/wwwroot/scripts/helper/createElement.ts [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    function setStyleIfApply() {
+        if (options.style) {
+            for (const key in options.style) {
+                if (options.style.hasOwnProperty(key)) {
+                    // @ts-ignore: Zuweisungen des style-Objekts sind dynamisch
+                    element.style[key] = options.style[key];
+                }
+            }
```
[the remark this answers is not in the document]
Comment (vanUbor) [reply WireMock/wwwroot/scripts/helper/createElement.ts]: sollte gelöst sein, bitte prüfe den fix, ich habe nicht deine vorgeschlagene Änderung vorgenommen sondern nutze Object.entries() stattdessen.

File: WireMock/wwwroot/scripts/index.ts [hunk 8]
```diff
@@ -115,47 +121,87 @@ function createMethods(item: Element, rawMap: any): void {
...
+    addMatcherCollectionButton.addEventListener("click", () => {
+        AddNewMapperCollectionToView(rawMap.Guid, boarder, headerDiv);
+        let mapJson = document.getElementById("rawMap-" + rawMap.Guid);
+        if (mapJson?.textContent) {
+            let rawMapContent = JSON.parse(mapJson.textContent);
+            rawMapContent.Request.Headers.push({Name: "N/A", IgnoreCase: false, Matchers: 
+                    [{Name: "WildcardMatcher", Pattern: "FooBar", IgnoreCase: false}]});
+            mapJson.textContent = JSON.stringify(rawMapContent, null, 1);
+        }
+    })
```
[the remark this answers is not in the document]
Comment (vanUbor) [reply WireMock/wwwroot/scripts/index.ts]: sollte gelöst sein, bitte prüfe den fix

File: WireMock/wwwroot/scripts/MatcherUI/MatcherUI.ts [hunk 9]
```diff
@@ -141,9 +148,61 @@ export default class MatcherUI {
...
+        }
+
+        // Optional: Hinzufügen von Event-Listenern für das Schließen
+        const closeButtons = errorModal?.querySelectorAll('[data-bs-dismiss="modal"]');
+        closeButtons?.forEach(button => {
+            button.addEventListener('click', () => {
+                this.closeErrorModal();
+            });
+        });
+    }
```
[the remark this answers is not in the document]
Comment (vanUbor) [reply WireMock/wwwroot/scripts/MatcherUI/MatcherUI.ts]: sollte gelöst sein, bitte prüfe den fix

File: WireMock/wwwroot/scripts/MatcherShells/HeaderMatcherShell.ts [hunk 10]
```diff
@@ -62,21 +71,60 @@ export default class HeaderMatcherShell extends MatcherShell {
...
+            const children = Array.from(headerparent!.children);
+            let headerIndex = children.indexOf(headerDiv!);
+            headerIndex -= 1
+            
+            rawMapContent.Request.Headers.splice(headerIndex, 1);
+
+            // write the object as json back to the text field
+            rawMap.textContent = JSON.stringify(rawMapContent, null, 1);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (vanUbor) [reply WireMock/wwwroot/scripts/MatcherShells/HeaderMatcherShell.ts]: ich habe die logik entfernt und nutze eine wiederverendbare lösung getHeaderIndex()

**Comment by vanUbor** [comment]
**Localization Issue**
Found some German comments and error messages in `getHeaderIndex.ts` that should be translated to English for consistency with the rest of the codebase:

  Lines to review:
  - L12: "Navigiere die angegebene Anzahl von Ebenen nach oben"
  - L20: `throw new Error('Konnte nicht ${levelsUp} Ebenen nach oben navigieren');`
  - L26: `throw new Error('Elternelement nicht gefunden');`

  Let's keep language consistent throughout the codebase for better collaboration.

**Comment by vanUbor** [comment]
**Code Refactoring Opportunity**
  The `MatcherShell` abstract class has been significantly improved, but there are still some methods like `renderMatcherHeader` that are quite large (around 40 lines).

  Consider extracting some of the component creation into smaller, reusable methods for better readability and maintainability. For example, the header name and ignore
  case sections could be separate methods.

**Comment by vanUbor** [comment]
Test Coverage
Great job adding helper functions like `createElement` and `getHeaderIndex`! These will definitely improve code maintainability.

  However, I don't see tests for these new utility functions. Since they're used throughout the UI, it would be valuable to add unit tests to ensure they behave as
  expected with different inputs and edge cases.

**Comment by vanUbor** [comment]
**Docker Improvements**
The Docker setup improvements are excellent - particularly adding Node.js for TypeScript compilation and proper volume permissions.

  For even better security, consider:
  1. Pinning specific versions for Node.js (e.g., `20.x.x` instead of just `20.x`)
  2. Adding a `.dockerignore` file to exclude unnecessary files from the build context
  3. Adding health checks to the docker-compose.yml file

**Comment by vanUbor** [comment]
**Constants Usage**
In several files like `cyberpunk-effects.js`, there are hardcoded values that would be better as named constants.

  For example, in functions like `addPulseAnimation()` and `createGlitchText()`, values like animation timing (2s), random thresholds (0.3), and glitch characters could
  be moved to constants at the top of the file for easier management and consistency.

**Comment by vanUbor** [comment]
## UI Cyberpunk Styling Review

The cyberpunk styling looks impressive! However, there are a few areas for improvement:

1. **Accessibility**: Some of the color contrast ratios might not meet WCAG standards, especially with the neon text colors. Consider testing with accessibility tools.

2. **Animations**: The glitch animations and audio effects are creative, but might negatively impact users with vestibular disorders or those who prefer reduced motion. Consider adding respect for  media query.

3. **Inline SVG**: Great use of SVG icons for buttons, but consider extracting them to a separate file or component for better maintainability.

**Comment by vanUbor** [comment]
## Docker Implementation Review

The Docker setup is mostly well implemented, but could benefit from a few improvements:

1. **Multi-stage build**: Excellent use of multi-stage builds to minimize image size.

2. **Security**: Good job moving to a non-root user, but consider these additional points:
   - Add a specific health check using  instruction
   - Consider using a more specific Node.js version pin (20.x is too broad)
   - Add  file to exclude test files, node_modules, etc.

3. **Environment Variables**: Consider adding more configurable environment variables rather than hardcoding values like ports.

**Comment by vanUbor** [comment]
## TypeScript Implementation Review

The TypeScript changes have several positive aspects, but also areas for improvement:

1. **Code Structure**: Good job using ES modules and implementing helper utilities.

2. **Type Safety**: 
   - In , the  utility is well-implemented, but consider adding stronger typing for the element properties.
   - Some event handlers could benefit from more specific event types rather than generic .

3. **Performance**: 
   - The DOM manipulation is efficient, using createElement rather than innerHTML for security and performance.
   - However, consider debouncing event handlers for input fields to prevent excessive processing.

4. **Target update to ES2023**: Good modernization of the TypeScript target, but ensure all browsers you support are compatible with this target.

**Comment by vanUbor** [comment]
## Specific Code Suggestions

### 1. In :



### 2. Dockerfile Improvement:

**Comment by vanUbor** [comment]
## Overall Review Summary

The PR introduces significant improvements across multiple areas:

### Strengths:
1. **UI Enhancements**: The cyberpunk styling is visually impressive and creates a unique user experience that fits well with the technical nature of the application.
2. **Docker Setup**: Well-structured Docker configuration with proper multi-stage builds and security considerations.
3. **Code Organization**: Good separation of concerns in both TypeScript and CSS files.

### Areas for Improvement:
1. **Accessibility**: Ensure color contrast meets WCAG standards and respect user preferences for reduced motion.
2. **Security**: Complete the Docker security configuration with additional best practices.
3. **Browser Compatibility**: Verify ES2023 compatibility with all supported browsers.

### Recommendation:
**Approve with suggestions** - The PR can be merged, but the team should address the accessibility and security concerns in follow-up work.

**Comment by vanUbor** [comment]
## Accessibility Improvements

Here's how to improve the cyberpunk styling for better accessibility:



Then add this to your layout:

**Comment by vanUbor** [comment]
## TypeScript Improvements

Here's a better implementation for the  file with improved typing:

name-

**Comment by vanUbor** [comment]
## Docker Configuration Improvements

Here are specific improvements for the Docker setup:

1. Add a  file to improve build performance:



2. Update Dockerfile with health check and security improvements:



3. Improve docker-compose.yml with resource limits and better security:

**Comment by vanUbor** [comment]
## JS Animation Improvements

Here's a better implementation for the cyberpunk-effects.js file with proper performance and accessibility:



This implementation:
1. Respects user preferences for reduced motion
2. Adds proper ARIA attributes for screen readers
3. Uses event delegation for better performance
4. Separates concerns into well-defined functions

**Comment by vanUbor** [comment]
## Final Review Summary

After detailed code review, here are the key recommendations:

### 1. Accessibility
- Add support for  media query
- Improve contrast ratios for better text readability
- Add keyboard navigation focus styles
- Include ARIA attributes for screen readers

### 2. TypeScript Improvements
- Use stronger typing with generics and interfaces
- Implement debouncing for input event handlers
- Improve element creation with proper typings
- Follow consistent naming conventions

### 3. Docker Security
- Add  file for build optimization
- Implement proper health checks
- Use specific version pins instead of major version ranges
- Add resource limits and proper restart policies
- Set correct file permissions and user restrictions

### 4. Performance
- Lazy-load audio files and other resources
- Use event delegation where applicable
- Minimize DOM manipulations
- Implement proper loading states

The cyberpunk UI design is impressive and adds character to the application, but implementation needs to balance aesthetics with accessibility and performance. With these improvements, the PR will be production-ready.

**Recommendation: Approve with changes requested.**

**Comment by vanUbor** [comment]
## Suggested code fix for TypeScript helper utility

Currently, the  helper lacks proper typing. Here's an improved implementation:



This version provides proper generic typing and better documentation.

**Review by vanUbor — COMMENTED** [review]
I've reviewed the PR in detail and left several code suggestions in the comments. While the cyberpunk UI styling and Docker improvements are impressive, there are some concerns around accessibility, performance, and security that should be addressed before final approval. See my specific comments for code-level suggestions.

**Comment by vanUbor** [comment]
I've implemented all the suggested changes in the PR review. The changes include:\n\n- Added support for prefers-reduced-motion media query and accessibility improvements\n- Improved Docker configuration with health checks, specific Node.js version, and security enhancements\n- Enhanced TypeScript typing with better interfaces and generics\n- Improved performance with event delegation and lazy loading\n\nAll issues have been addressed. Closing this thread.

**Comment by vanUbor** [comment]
I've implemented all the suggested changes from the PR review. Here's a summary of the improvements:\n\n## Accessibility Improvements\n- Added prefers-reduced-motion media query support\n- Added skip-to-content link for keyboard navigation\n- Improved focus styles for better accessibility\n- Added ARIA attributes for screen readers\n- Created .sr-only utility class for screen reader text\n\n## Docker Improvements\n- Added HEALTHCHECK to the Dockerfile\n- Pinned Node.js to a specific version (20.11.1) for better security\n- Added resource limits and security options in docker-compose.yml\n- Added configurable environment variables\n- Improved user handling with proper permissions\n\n## TypeScript Improvements\n- Created an ElementCreateOptions interface\n- Added proper JSDoc documentation\n- Improved element creation with generic typing\n- Consolidated code for better maintainability\n\n## Performance Improvements\n- Used event delegation for better event handling\n- Implemented lazy loading for audio resources\n- Added debouncing to prevent excessive processing\n- Used batch operations to reduce DOM reflows\n\nAll issues have been addressed, and the code is now more accessible, secure, and maintainable.

**Comment by vanUbor** [comment]
All suggested improvements have been implemented and the code is ready for final review and merge. I've made sure to test all the changes, and they're working as expected.
